src/pid/AbstractController.java

```diff
@@ -0,0 +1,34 @@
+package pid;
+
+import pid.interfaces.Controller;
+
+public abstract class AbstractController implements Controller {
+  protected PIDFTerms pidfTerms;
+  protected double setpoint;
+
+  public AbstractController(double kP, double kI, double kD, double kF) {
```

Comment:
You can add a ctor that gets `PIDFTerms` as a parameter and call it from here. That would also enable you to call it from inheritors. Genrally, I think they should create the PIDFTerms outside to make them more meaningful, and to avoid duplicate naming (`driveTrainTerms` instead of `driveTrainP`, `driveTrainI`, `driveTrainD`, `driveTrainF`).

Reply:
Done